<a href="https://colab.research.google.com/github/adgcal/Anjan_collab_R/blob/main/3_bioinfo_R.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# R Curiosity :Bioinformatics 

## Extracting data from PUBMED 


- PubMed is an online repository of references and abstracts of publications in the fields of medicine and life sciences. 
- PubMed is a free resource that is developed and maintained by the National Center for Biotechnology Information (NCBI)
- Alternatively, PubMed can be programmatically queried via the NCBI Entrez E-utilities interface.


In [ ]:
install.packages("devtools")

In [ ]:
library(devtools)

Loading required package: usethis



In [ ]:
#installed.packages()

[link text](https://)library(easyPubMed)
# easyPubmed description
- easyPubMed is an R interface to the Entrez Programming Utilities aimed at allowing an easy and smooth programmatic access to PubMed. 
- The package is suitable for batch downloading large volumes of records (via the batch_pubmed_download() function), and comes with a set of functions to perform basic processing of the PubMed query output.
- easyPubMed can request and handle PubMed records in either XML or TXT format. 

# A simple example 
- easyPubMed
- rentrez 


In [ ]:
install_github("dami82/easyPubMed")

In [ ]:
install.packages("rentrez")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependency ‘XML’




In [ ]:
library(easyPubMed)
my_query <- 'doxyrubicin'
my_entrez_id <- get_pubmed_ids(my_query)
my_abstracts_txt <- fetch_pubmed_data(my_entrez_id, format = "abstract")
head(my_abstracts_txt)

[1] "1. ACS Appl Mater Interfaces. 2021 Mar 3;13(8):9316-9328. doi: "          
[2] "10.1021/acsami.0c14730. Epub 2020 Oct 22."                                
[3] ""                                                                         
[4] "Aptamer-Functionalized Upconverting Nanoformulations for Light-Switching "
[5] "Cancer-Specific Recognition and In Situ Photodynamic-Chemo Sequential "   
[6] "Theranostics."

In [ ]:

# Define the query
my_query <- "doxorubicin"

# Search for articles related to the query and retrieve their IDs
search_results <- entrez_search(db = "pubmed", term = my_query)
# Number of search results 
length(search_results$ids)
my_entrez_id <- search_results$ids
my_abstracts_txt <- entrez_fetch(db = "pubmed", id = my_entrez_id, rettype = "abstract", retmode = "text")
# Read the abstracts file into a character vector
writeLines(my_abstracts_txt, "my_abstracts.txt")


[1] 20

In [ ]:
# Define the query
my_query <- "side effects of COVID vaccine from pfizer"

# Search for articles related to the query and retrieve their IDs
search_results <- entrez_search(db = "pubmed", term = my_query)
# Number of search results 
length(search_results$ids)
my_entrez_id <- search_results$ids
my_abstracts_txt <- entrez_fetch(db = "pubmed", id = my_entrez_id, rettype = "abstract", retmode = "text")
# Read the abstracts file into a character vector
writeLines(my_abstracts_txt, "my_pfizer.txt")

[1] 20

In [ ]:
writeLines(my_abstracts_txt, "my_abstracts.txt")


# From *in vivo* , *in vitro* to *in silico*

Bioinformatics came into existence as a result of cumulative efforts of many scientists. But if we have to give the credit to two individuals for setting up the basic platform for bioinformatics we may name Margartette Dayhoff (who discovered consensus sequence) G.N.Ramachandran (who introduced the widely used phi-psi plot) .In the next stage of development , we may name Needleman and Wunch (who introduced global alignment, and a little modification of their dynamical programming algorithm leads to  local alignment). We must mention Atschachul who introduced blast search. A phenomenal growth of the subject was seen, once whole data for genome sequencing started pouring in literature and crystal structure data expanded like never before (thanks to the Cambridge crystallographic repository).  




# The *seqinr* package

In [ ]:
install.packages("seqinr")

Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

also installing the dependencies ‘pixmap’, ‘sp’, ‘Rcpp’, ‘RcppArmadillo’, ‘ade4’, ‘segmented’




In [ ]:
library(seqinr)

## Dengue

- Dengue is a viral disease transmitted by mosquitoes, mainly the Aedes aegypti mosquito.
- It is caused by the Dengue virus, which is a single-stranded RNA virus belonging to the Flaviviridae family.
- Dengue is prevalent in tropical and subtropical regions of the world, and its symptoms include high fever, severe headache, joint and muscle pain, and a rash. 
- There is no specific treatment for Dengue, and supportive care is the mainstay of management.

## H1N1:
- H1N1 is a subtype of the Influenza A virus that causes seasonal flu in humans.
- It is a segmented RNA virus belonging to the Orthomyxoviridae family, and it can infect humans, pigs, and birds.
- H1N1 caused a pandemic in 2009-2010, also known as the swine flu pandemic.
- Symptoms of H1N1 include fever, cough, sore throat, body aches, and fatigue.
- Vaccines are available to prevent H1N1, and antiviral drugs can be used to treat severe cases.

A function to access sequence from NCBI

In [ ]:
getncbiseq <- function(accession)
{
require("seqinr") # this function requires the SeqinR R package
# first find which ACNUC database the accession is stored in:
dbs <- c("genbank","refseq","refseqViruses","bacterial")
numdbs <- length(dbs)
for (i in 1:numdbs)
      {
db <- dbs[i]
choosebank(db)
# check if the sequence is in ACNUC database 'db':
resquery <- try(query(".tmpquery", paste("AC=", accession)), silent = TRUE)

if (!(inherits(resquery, "try-error"))) {
  queryname <- "query2"
  thequery <- paste("AC=", accession, sep="")
  query2 <- query(queryname, thequery)
  # see if a sequence was retrieved:
  seq <- getSequence(query2$req[[1]])
  closebank()
  return(seq)
}
closebank()
}
print(paste("ERROR: accession",accession,"was not found"))
}

# The Dengue virus sequence 


The variable dengueseq is a vector containing the nucleotide sequence. Each element of the vector contains one nucleotide of the sequence. Therefore, to print out a certain subsequence of the sequence, we just need to type the name of the vector dengueseq followed by the square brackets containing the indices for those nucleotides. For example, the following command prints out the first 50 nucleotides of the DEN-1 Dengue virus genome sequence

In [ ]:
dengueseq <- getncbiseq("NC_001477")

# First 20 members of the  Dengue virus sequence 


In [ ]:
dengueseq[1:20]


[1] "a" "g" "t" "t" "g" "t" "t" "a" "g" "t" "c" "t" "a" "c" "g" "t" "g" "g" "a"
[20] "c"

In [ ]:
write.fasta(names="/content/sample_data/DEN-1", sequences=dengueseq, file.out="den1.fasta")


In [ ]:
dengue <- read.fasta(file = "/content/den1.fasta")


Length of a DNA sequence
Once you have retrieved a DNA sequence, we can obtain some simple statistics to describe that sequence, such as the sequence’s total length in nucleotides. In the above example, we retrieved the DEN-1 Dengue virus genome se- quence, and stored it in the vector variable dengueseq. To subsequently obtain the length of the genome sequence, we would use the length() function, typing:


In [ ]:
dengueseq <- dengue[[1]]
length(dengueseq)

[1] 10735

 ## Base composition of a DNA sequence


In [ ]:
table(dengueseq)

dengueseq
   a    c    g    t 
3426 2240 2770 2299 

# H1N1 Sequence

In [ ]:
h1n1seq<-getncbiseq("HM624086")	


In [ ]:
write.fasta(names="/content/sample_data/h1n1", sequences=h1n1seq, file.out="h1n1.fasta")


In [ ]:
h1n1<-read.fasta(file = "/content/h1n1.fasta")


In [ ]:
table(h1n1)

/content/sample_data/h1n1
  a   c   g   t 
603 314 377 407 

In [ ]:
h1n1seq<-h1n1[[1]]

# GC contents 

In [ ]:
GC(dengueseq)
GC(h1n1seq)


[1] 0.4666977

[1] 0.4062316

# Counting (alternative of Table) 

1.   List item
2.   List item



In [ ]:
count(dengueseq,1)
count(h1n1seq,1)


   a    c    g    t 
3426 2240 2770 2299 


  a   c   g   t 
603 314 377 407 

# Sequence alignment

In [ ]:
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")


Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)



In [ ]:
BiocManager::install("Biostrings")



In [ ]:
library(BiocManager)


In [ ]:
#BiocManager::install("msa")


In [ ]:
library(Biostrings)
#library(msa)


In [ ]:
help.search("clustal", package = "msa")


In [ ]:
seq1<- read.fasta("den1.fasta")
seq2<- read.fasta("h1n1.fasta")

In [ ]:
# Perform Clustal alignment
# Two example DNA sequences

sseq1 <- "ATCGCGATCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGC"
sseq2 <- "ATCGCGATCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCGCAG"

alignment <- msa(c(sseq1, sseq2), method = "ClustalW", type = "dna")
alignment_matrix <- as.matrix(alignment)

# Write the matrix to a file
write.table(alignment_matrix, file = "my_alignment.txt", sep = "\t", quote = FALSE, row.names = FALSE, col.names = FALSE)



## RNA-seq analysis in R


<https://combine-australia.github.io/RNAseq-R/00-r-rstudio-intro.html>



Mouse mammary data (counts): https://figshare.com/s/1d788fd384d33e913a2a </p>
Drosophila data (counts): https://figshare.com/s/e08e71c42f118dbe8be6

In [ ]:
list.files()

[1] "den1.fasta"        "h1n1.fasta"        "mouse_c2_v5.rdata"
[4] "sample_data"

Smarter installationof Bioconductor package 
<https://bioconductor.org/packages/release/bioc/>



** Example 1: Biobase installation**

In [ ]:
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager") 
BiocManager::install("Biobase")

In [ ]:
library(Biobase)

In [ ]:
BiocManager::install(c("GenomicFeatures", "AnnotationDbi"))

In [ ]:
library(AnnotationDbi)

# Problem 
How many of each of the four nucleotides A, C, T and G, and any other symbols, are there in the Mycobacterium leprae Note: other symbols apart from the four nucleotides A/C/T/G may appear in a sequence. They correspond
to positions in the sequence that are are not clearly one base or another and they are due, for example,
to sequencing uncertainties. or example, the symbol ‘N’ means ‘aNy base’, while ‘R’ means ‘A or G’
(puRine). There is a table of symbols at www.bioinformatics.org/sms/iupac.html.

In [ ]:
acc<- 'NC_002677' 


https://www.ncbi.nlm.nih.gov/nuccore/NC_000962.2?report=fasta

URL=<https://www.ncbi.nlm.nih.gov/nuccore/NC_002677.1/?report=genbank&log$=nuclalign&from=1&to=3414141>

# Quiz Bioinfo 
 - What information about the rabies virus sequence (NCBI accession NC_001542) can you obtain from its annotations.
What does it say in the DEFINITION and ORGANISM fields of its NCBI record? Note: rabies virus is the virus responsible for rabies, which is classified by the WHO as a neglected tropical disease.

- How many nucleotide sequences are there from the bacterium Chlamydia trachomatis in the NCBI Sequence Database? Note: the bacterium Chlamydia trachomatis is responsible for causing trachoma, which is classified by the WHO as a neglected tropical disease.
- How many nucleotide sequences are there from the bacterium Chlamydia trachomatis in the RefSeq part of the NCBI Sequence Database?
- How many nucleotide sequences were submitted to NCBI by Matthew Berriman?
- How many nucleotide sequences from nematode worms are there in the RefSeq Database? Note that several parasitic nematode worms cause neglected tropical diseases, including Brugia malayi and Wucheria bancrofti, which cause lymphatic filariasis; Loa loa, which causes subcutaneous filariasis; Onchocerca volvulus, which causes onchocerciasis; and Necator americanus, which causes soil-transmitted helminthiasis.
- How many nucleotide sequences for collagen genes from nematode worms are there in the NCBI Database?
- How many mRNA sequences for collagen genes from nematode worms are there in the NCBI Database?
- How many protein sequences for collagen proteins from nematode worms are there in the NCBI database?
- What is the accession number for the Trypanosoma cruzi genome in NCBI? Do you see genome se- quences for more than one strain of Trypanosoma cruzi? Note that the Trypanosoma cruzi causes Chagas disease, which is classified as a neglected tropical disease by the WHO.
-How many fully sequenced nematode worm species are represented in the NCBI Genome database?

# Answers




a) What information about the rabies virus sequence (NCBI accession NC_001542) can you obtain from its annotations?

From the annotations of the rabies virus sequence (NCBI accession NC_001542), we can obtain various information such as its DEFINITION, which describes the source organism, length, and molecular type of the sequence. We can also find information about the ORGANISM, which is "Rabies virus" in this case. Additionally, annotations can provide information about the features of the sequence, such as genes, coding regions, and other relevant information.

b) How many nucleotide sequences are there from the bacterium Chlamydia trachomatis in the NCBI Sequence Database?

*See code below*.


c) How many nucleotide sequences are there from the bacterium Chlamydia trachomatis in the RefSeq part of the NCBI Sequence Database?

There are 379 nucleotide sequences from the bacterium Chlamydia trachomatis in the RefSeq part of the NCBI Sequence Database.

d) How many nucleotide sequences were submitted to NCBI by Matthew Berriman?

*See code below*.

e) How many nucleotide sequences from nematode worms are there in the RefSeq Database?

*See code below*.


f) How many nucleotide sequences for collagen genes from nematode worms are there in the NCBI Database?

*See code below*.


g) How many mRNA sequences for collagen genes from nematode worms are there in the NCBI Database?

*See code below*.


h) How many protein sequences for collagen proteins from nematode worms are there in the NCBI database?

*See code below*.


i) What is the accession number for the Trypanosoma cruzi genome in NCBI? Do you see genome sequences for more than one strain of Trypanosoma cruzi?

The accession number for the Trypanosoma cruzi genome in NCBI is GCF_000002325.2. Yes, there are genome sequences available for multiple strains of Trypanosoma cruzi in NCBI.

j) How many fully sequenced nematode worm species are represented in the NCBI Genome database?

As of my knowledge cutoff date of September 2021, there were 68 fully sequenced nematode worm species represented in the NCBI Genome database.

In [ ]:
# a
# Load the required packages
library(rentrez)
library(Biostrings)

# Set the NCBI accession number for the Rabies virus genome
acc <- "NC_001542"
rabbies<- getncbiseq(acc)
write.fasta(names="/content/sample_data/rabbies", sequences=rabbies, file.out="rabbies.fasta")


- Visit <https://www.ncbi.nlm.nih.gov/nuccore/NC_001542.1>
- It indicates that the gene "GeneID:1489853" is assigned the locus tag "RABVgp1". 
-  CDS   5418..11846

In [ ]:
#b
# Load necessary packages
library(rentrez)

# Search NCBI Sequence Database for Chlamydia trachomatis nucleotide sequences
term <- "Chlamydia trachomatis"
search <- entrez_search(db="nuccore", term=term, retmax=0)
search$count
search$QueryTranslation

[1] 106051

[1] "\"Chlamydia trachomatis\"[Organism] OR Chlamydia trachomatis[All Fields]"

In [ ]:
# c
library(rentrez)
search <- entrez_search(db="nucleotide", term="Chlamydia trachomatis[Organism] AND refseq[filter]", retmax=0)
count <- as.integer(search$count)
cat("Number of nucleotide sequences from Chlamydia trachomatis in the RefSeq database:", count)


Number of nucleotide sequences from Chlamydia trachomatis in the RefSeq database: 379

In [ ]:
# d
library(rentrez)

# Search NCBI Nucleotide Database for sequences submitted by Matthew Berriman
term <- "Berriman M[Author]"
search <- entrez_search(db="nuccore", term=term, retmax=0)
count <- search$count

# Print the result
cat("Number of nucleotide sequences submitted by Matthew Berriman:", count, "\n")


Number of nucleotide sequences submitted by Matthew Berriman: 648267 


In [ ]:
#e
library(rentrez)

# Search for nucleotide sequences from nematode worms in RefSeq Database
term <- "nematoda[Organism] AND refseq[filter]"
search <- entrez_search(db="nuccore", term=term, retmax=0)

# Get the count of nucleotide sequences
count <- search$count
print(paste("Number of nucleotide sequences from nematode worms in RefSeq Database:", count))


[1] "Number of nucleotide sequences from nematode worms in RefSeq Database: 205389"


In [ ]:
#f
library(rentrez)

# Search NCBI Nucleotide Database for collagen genes from nematode worms
term <- "collagen AND nematode"
search <- entrez_search(db="nucleotide", term=term, retmax=0)
cat("Number of nucleotide sequences for collagen genes from nematode worms in NCBI Database:", search$count)


Number of nucleotide sequences for collagen genes from nematode worms in NCBI Database: 10735

In [ ]:
#g
library(rentrez)

# Set search parameters
term <- "nematoda[orgn] AND collagen AND biomol_mrna[prop]"
search <- entrez_search(db="nucleotide", term=term, retmax=0)

# Print the count result
cat("Number of mRNA sequences for collagen genes from nematode worms in NCBI database:", search$count)


Number of mRNA sequences for collagen genes from nematode worms in NCBI database: 7606

In [ ]:
#h
library(rentrez)

# Search NCBI Protein Database for collagen proteins from nematode worms
term <- "nematode collagen protein"
search <- entrez_search(db="protein", term=term, retmax=0)

# Print the number of protein sequences
cat("Number of protein sequences for collagen proteins from nematode worms:",
    format(search$count, big.mark=",", scientific=FALSE))



Number of protein sequences for collagen proteins from nematode worms: 10,791

In [ ]:
#i 
library(rentrez)

# Search for the Trypanosoma cruzi genome in the nucleotide database
term <- "Trypanosoma cruzi[Organism]"
search <- entrez_search(db = "nucleotide", term = term)

# Print the accession numbers for the results
if (search$count > 0) {
  res <- entrez_fetch(db = "nucleotide", id = search$ids[1], rettype = "acc")
  cat("Accession number for the T. cruzi genome:", res, "\n")
} else {
  cat("No results found.\n")
}
# ANSWER IS WRONG

Accession number for the T. cruzi genome: OP356744.1
 


In [ ]:
library(rentrez)

# Search for all T. cruzi sequences in the nucleotide database
term <- "Trypanosoma"
search <- entrez_search(db = "nucleotide", term = term)

# Filter the results to include only genome sequences
ids <- search$ids[search$ds == "nuccore" & grepl("genome", search$description)]
if (length(ids) > 0) {
  # Print the accession numbers for the genome sequences
  res <- entrez_fetch(db = "nucleotide", id = ids, rettype = "acc")
  cat("Accession numbers for T. cruzi genome sequences:\n", res, "\n")
} else {
  cat("No genome sequences found.\n")
}


No genome sequences found.


# Added Note 
RefSeq (Reference Sequence) is a comprehensive, curated, non-redundant, and well-annotated database of reference sequences for a variety of organisms, including bacteria, viruses, fungi, plants, and animals. The RefSeq database provides a stable reference for genome annotation, gene and protein identification, transcriptomic analysis, and other biological investigations. RefSeq entries are derived from multiple sources, including genomic, transcriptomic, and protein sequence data, and are reviewed and annotated by experts to ensure high quality and accuracy. RefSeq records are regularly updated and maintained, and are widely used as a standard reference for a variety of biological research applications. Thus the number of entries (379)<< (106051)